### ISSUE TITLE
Applying the filter method to the FASTQ file results in empty filtered file

### ISSUE THREAD

**Issue by alexvasilikop** [blobtoolkit/blobtoolkit#49]
Hi,

I used the interactive version to make my custom filtering and then save the list (full json to use it for filtering the fasta assembly and the reads separately). Filtering for the assembly completed without problems but when I provide the FASTQ (forward reads) there is no error printed on the screen but the output FASTQ (filtered) is empty.

This is the command I used:
./blobtoolkit/blobtools2/blobtools filter --json Rs_adtrimmed_BlobDir.gc-len-cov-manual.json \
--output /blobtoolkit/Rsadtrimmed_filtered_BlobDir \
--fastq ../BXQ_LOSDE_1_1_HL35GDRXY.UDI319_clean.adap_trimmed.fastq  \
--cov DIAMOND_BLAST_AD_TRIMMED/aln-rotaria-masurca-ad-trimmed-pe.sorted.bam \ 
blobtoolkit/Rs_adtrimmed_BlobDir

Any idea what the problem is?

**Comment by caioald** [comment]
I am facing the same problem here, any news on this?
When I filter the long reads, it usually works., but when I filter the short reads, I get empty files.

`blobtools filter --param buscoregions_phylum--Keys=Proteobacteria \
  --param buscoregions_phylum--Keys=Actinobacteria \
  --param buscoregions_superkingdom--Keys=Bacteria \
  --fastq /home/Ac2012_1_IS350_Clean_1.fastq \
  --fastq /home/Ac2012_1_IS350_Clean_2.fastq \
  --cov ./minimap/Ac2012_1.Ac2012_1_IS350_Clean.bam --summary filter_buscoregion.json \
  --output ./filter_buscoregion \
  ./blobtools/Ac2012_1`

**Comment by rjchallis** [comment]
Sorry for letting this issue get stale. This looks like an issue with the way blobtools is using pysam to process the bam files. Different minimap settings (e.g. for short vs long reads) affect the output format and some combinations of BAM format/pysam version seem to be causing problems for blobtools. I think I will need to refactor the code here to use an alternate approach.

**Comment by rjchallis** [comment]
In the meantime, you should be able to achieve the filtering with a combination of blobtools, bedtools and seqtk by using blobtools filter --table to get a list of sequence IDs, bedtools/grep to get the corresponding read IDs and seqtk to filter the fastq file, e.g.:
 
```
blobtools filter --param buscoregions_phylum--Keys=Proteobacteria \
    --param buscoregions_phylum--Keys=Actinobacteria \
    --param buscoregions_superkingdom--Keys=Bacteria \
    --table table.tsv ilSynVesp1_1.1

bedtools bamtobed -i ./minimap/Ac2012_1.Ac2012_1_IS350_Clean.bam | grep -wFf <(tail -n +2 table.tsv | cut -f 2) table.tsv | cut -f 4 > name.lst

seqtk subseq /home/Ac2012_1_IS350_Clean_1.fastq name.lst > out_1.fq
seqtk subseq /home/Ac2012_1_IS350_Clean_2.fastq name.lst > out_2.fq
```

**Comment by caioald** [comment]
Hi Richard,

I did what you suggested but I had to remove the table.tsv from the grep part (... `grep -wFf <(tail -n +2 table.tsv | cut -f 2) | cut -f 4 > name.lst`

However, I didn't get the same number of reads from both sides:
`grep -c "/1" name.lst` = 18690664
`grep -c "/2" name.lst` = 18609328

So the *.fq files were assimetric.

Do you have any thoughts about this?

Thanks again

**Comment by caioald** [comment]
Would this be because only one of the reads in the pair was mapped to the reference?
How could I overcome this?

**Comment by rjchallis** [comment]
yes, this must be due to only one read mapping. To get both pairs you could put all of the read prefixes (minus the /1 or /2 into a file then use sort -u to get a single list with all read prefixes and use this to make 2 new list files for seqtk, one with the /1 suffix and one with the /2 suffix

**Comment by Sabrin2020** [comment]
@rjchallis can you please break the grep part of the code here to more simple one when I try it as it is to my data i get an empty file at the end and been stuck for a while
`/minimap/Ac2012_1.Ac2012_1_IS350_Clean.bam **| grep -wFf <(tail -n +2 table.tsv | cut -f 2) table.tsv | cut -f 4** > name.lst`

**Comment by caioald** [comment]
Hi @Sabrin2020,
This was the successful bedtools code for me:

`bedtools bamtobed -i ./minimap/Ac2012_1.Ac2012_1_IS350_Clean.bam | grep -wFf <(tail -n +2 table.tsv | cut -f 2) | cut -f 4 > name.lst`

**Comment by rjchallis** [comment]
@Sabrin2020, if the command from @caioald can you let me know how the outputs of the sub-commands look with
```
tail -n +2 table.tsv | cut -f 2 | head
```
and
```
bedtools bamtobed -i ./minimap/Ac2012_1.Ac2012_1_IS350_Clean.bam | head
```

**Comment by Sabrin2020** [comment]
> @Sabrin2020, if the command from @caioald doesn't work can you let me know how the outputs of the sub-commands look with
> 
> ```
> tail -n +2 table.tsv | cut -f 2 | head
> ```
> 
> and
> 
> ```
> bedtools bamtobed -i ./minimap/Ac2012_1.Ac2012_1_IS350_Clean.bam | head
> ```

```
tail -n +2 no-Chordata.tsv | cut -f 2 | head
5
6
9
11
12
13
15
17
18
20
```
```
bedtools bamtobed -i /user/asga9989/Gills/2_blobtools/bam.set1set2set3.rest/18424nigboc_mapped.bam |head
13      0       136     K00149:35:H7YWGBBXX:3:1202:26707:42530/2        60      +
13      0       117     K00149:35:H7YWGBBXX:3:1215:28158:25070/1        60      +
13      0       135     K00149:35:H7YWGBBXX:3:1218:18893:14150/1        60      +
13      0       135     K00149:35:H7YWGBBXX:3:1218:19076:14221/1        60      +
13      0       135     K00149:35:H7YWGBBXX:3:1218:19004:14308/1        60      +
13      0       135     K00149:35:H7YWGBBXX:3:1218:19329:14308/1        60      +
13      0       135     K00149:35:H7YWGBBXX:3:1218:19461:14607/1        60      +
13      0       135     K00149:35:H7YWGBBXX:3:1218:18994:14643/1        60      +
13      0       135     K00149:35:H7YWGBBXX:3:1218:19319:14713/1        60      +
13      0       149     K00149:35:H7YWGBBXX:3:2211:31182:18036/1        60      +
```

**Comment by Sabrin2020** [comment]
@rjchallis So, blobdirs never seem to contains any read ids. The sequence identifiers in e.g. identifiers.json but also all the bestsumorder files **always refer to the reference assembly only**. Read ids we can only get from the bam files, by cross-referencing them with the list of contigs provided by blobtools, **Would you please confirm on that?**
